Typescript 4.8.2 support

Typescript version 4.8.2 just got released. There is deprecated function used in the plugin.

Functions:
**createParameterDeclaration**

Error:
```
TypeError: Invalid arguments
    at Object.createParameterDeclaration (/usr/local/lib/node_modules/@domain/protoc-gen-ts/node_modules/typescript/lib/typescript.js:170461:19)
    at createParameter (/usr/local/lib/node_modules/@domain/protoc-gen-ts/src/rpc.js:154:23)
    at Object.createGrpcInterfaceType (/usr/local/lib/node_modules/@domain/protoc-gen-ts/src/rpc.js:161:31)
    at Object.<anonymous> (/usr/local/lib/node_modules/@domain/protoc-gen-ts/src/index.js:54:32)
    at Module._compile (node:internal/modules/cjs/loader:1126:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1180:10)
    at Module.load (node:internal/modules/cjs/loader:1004:32)
    at Function.Module._load (node:internal/modules/cjs/loader:839:12)
    at Module.require (node:internal/modules/cjs/loader:1028:19)
    at require (node:internal/modules/cjs/helpers:102:18)
--ts_out: protoc-gen-ts: Plugin failed with status code 1.
```

According the line 170461 in typescript package it seems like the function was removed completly.
```
// Deprecations fall into one of three categories:
//
// - "soft" - Soft deprecations are indicated with the `@deprecated` JSDoc Tag.
// - "warn" - Warning deprecations are indicated with the `@deprecated` JSDoc Tag and a diagnostic message (assuming a compatible host).
// - "error" - Error deprecations are either indicated with the `@deprecated` JSDoc tag and will throw a `TypeError` when invoked, or removed from the API entirely.
```

@puzzle9 that also doesn't work. I've forked the repo [here](https://github.com/long1eu/protoc-gen-ts) and this is what I get
```
node:internal/modules/cjs/loader:988
  throw err;
  ^

Error: Cannot find module '../src/index'
Require stack:
- /usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
    at Function.Module._load (node:internal/modules/cjs/loader:833:27)
    at Module.require (node:internal/modules/cjs/loader:1057:19)
    at require (node:internal/modules/cjs/helpers:103:18)
    at Object.<anonymous> (/usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js:2:1)
    at Module._compile (node:internal/modules/cjs/loader:1155:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
    at Module.load (node:internal/modules/cjs/loader:1033:32)
    at Function.Module._load (node:internal/modules/cjs/loader:868:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [ '/usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js' ]
}
--ts_out: protoc-gen-ts: Plugin failed with status code 1.
```

> @puzzle9 that also doesn't work. I've forked the repo [here](https://github.com/long1eu/protoc-gen-ts) and this is what I get
> 
> ```
> node:internal/modules/cjs/loader:988
>   throw err;
>   ^
> 
> Error: Cannot find module '../src/index'
> Require stack:
> - /usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js
>     at Function.Module._resolveFilename (node:internal/modules/cjs/loader:985:15)
>     at Function.Module._load (node:internal/modules/cjs/loader:833:27)
>     at Module.require (node:internal/modules/cjs/loader:1057:19)
>     at require (node:internal/modules/cjs/helpers:103:18)
>     at Object.<anonymous> (/usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js:2:1)
>     at Module._compile (node:internal/modules/cjs/loader:1155:14)
>     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1209:10)
>     at Module.load (node:internal/modules/cjs/loader:1033:32)
>     at Function.Module._load (node:internal/modules/cjs/loader:868:12)
>     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:81:12) {
>   code: 'MODULE_NOT_FOUND',
>   requireStack: [ '/usr/local/lib/node_modules/protoc-gen-ts/bin/protoc-gen-ts.js' ]
> }
> --ts_out: protoc-gen-ts: Plugin failed with status code 1.
> ```

yea

because it is the source code that needs to be compiled before it can be used

simple you can try

```sh
mkdir protoc-gen-ts
cd protoc-gen-ts
yarn add protoc-gen-ts
mv node_modules/protoc-gen-ts ./pgt
```

`cd pgt`

change `package.json` `"typescript": "^4.3.4"` to `"typescript": "4.7.4"`

```sh
yarn
mv bin/protoc-gen-ts.js bin/protoc-gen-ts
```

remove global `protoc-gen-ts` and add path this file

try

--- a/examples/pure/package.json
+++ b/examples/pure/package.json
@@ -11,6 +11,6 @@
     },
     "devDependencies": {
         "protoc-gen-ts": "file:../../bazel-bin/package",
-        "typescript": "^4.2.4"
+        "typescript": "^4.9.3"
     }
 }

--- a/package.json
+++ b/package.json
@@ -26,7 +26,7 @@
   },
   "peerDependencies": {
     "google-protobuf": "^3.13.0",
-    "typescript": ">=4"
+    "typescript": ">=4.9.3"
   },
   "devDependencies": {
     "@bazel/bazelisk": "^1.9.0",
@@ -35,18 +35,18 @@
     "@bazel/rollup": "5.7.1",
     "@bazel/cypress": "5.7.1",
     "cypress": "7.3.0",
-    "@grpc/grpc-js": "^1.3.4",
+    "@grpc/grpc-js": "^1.7.3",
     "google-protobuf": "^3.19.1",
     "grpc-web": "^1.4.2",
     "jasmine": "^3.7.0",
     "jasmine-core": "^3.7.1",
-    "typescript": "^4.3.4",
+    "typescript": "^4.9.3",
     "@rollup/plugin-commonjs": "^23.0.2",
     "@rollup/plugin-node-resolve": "^15.0.1",
     "rollup": "^3.3.0",
     "http-server": "^14.1.1",
     "@types/google-protobuf": "^3.15.5",
     "@types/jasmine": "^3.7.7",
-    "@types/node": "^15.12.5"
+    "@types/node": "^18.11.9"
   }
 }

--- a/src/descriptor.ts
+++ b/src/descriptor.ts
@@ -36,7 +36,7 @@ export function createEnum(
   enumDescriptor: descriptor.EnumDescriptorProto,
   parentName: string = ''
 ): ts.EnumDeclaration {
-  const values = [];
+  const values: ts.EnumMember[] = [];
 
   for (const valueDescriptor of enumDescriptor.value) {
     values.push(
@@ -51,7 +51,6 @@ export function createEnum(
   }
   return comment.addDeprecatedJsDoc(
     ts.factory.createEnumDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
       ts.factory.createIdentifier(`${parentName}${enumDescriptor.name}`),
       values,
@@ -137,7 +136,6 @@ function createFromObject(
                 ts.factory.createParameterDeclaration(
                   undefined,
                   undefined,
-                  undefined,
                   ts.factory.createArrayBindingPattern([
                     ts.factory.createBindingElement(
                       undefined,
@@ -171,7 +169,6 @@ function createFromObject(
           undefined,
           [
             ts.factory.createParameterDeclaration(
-              undefined,
               undefined,
               undefined,
               "item",
@@ -289,7 +286,6 @@ function createFromObject(
   statements.push(ts.factory.createReturnStatement(messageIdentifier));
 
   return ts.factory.createMethodDeclaration(
-    undefined,
     [ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)],
     undefined,
     ts.factory.createIdentifier("fromObject"),
@@ -297,7 +293,6 @@ function createFromObject(
     undefined,
     [
       ts.factory.createParameterDeclaration(
-        undefined,
         undefined,
         undefined,
         dataIdentifier,
@@ -355,7 +350,6 @@ function createToObject(
                 ts.factory.createParameterDeclaration(
                   undefined,
                   undefined,
-                  undefined,
                   ts.factory.createArrayBindingPattern([
                     ts.factory.createBindingElement(
                       undefined,
@@ -404,7 +398,6 @@ function createToObject(
           undefined,
           [
             ts.factory.createParameterDeclaration(
-              undefined,
               undefined,
               undefined,
               "item",
@@ -502,7 +495,6 @@ function createToObject(
   return ts.factory.createMethodDeclaration(
     undefined,
     undefined,
-    undefined,
     ts.factory.createIdentifier("toObject"),
     undefined,
     undefined,
@@ -523,7 +515,6 @@ export function createNamespace(
 
   for (let i = identifiers.length - 1; i >= 0; i--) {
     decl = ts.factory.createModuleDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
       ts.factory.createIdentifier(identifiers[i]),
       decl as ts.ModuleBlock,
@@ -643,7 +634,7 @@ function createPrimitiveMessageSignature(
     if (field.isMap(fieldDescriptor)) {
       const [keyDescriptor, valueDescriptor] = type.getMapDescriptor(
         fieldDescriptor.type_name,
-      ).field;
+      )!.field;
 
       let valueType = field.getType(valueDescriptor, rootDescriptor);
 
@@ -654,10 +645,8 @@ function createPrimitiveMessageSignature(
       fieldType = ts.factory.createTypeLiteralNode([
         ts.factory.createIndexSignature(
           undefined,
-          undefined,
           [
             ts.factory.createParameterDeclaration(
-              undefined,
               undefined,
               undefined,
               "key",
@@ -875,10 +864,8 @@ function createConstructor(
 
   return ts.factory.createConstructorDeclaration(
     undefined,
-    undefined,
     [
       ts.factory.createParameterDeclaration(
-        undefined,
         undefined,
         undefined,
         dataIdentifier,
@@ -919,7 +906,6 @@ function createGetter(
 
   return comment.addDeprecatedJsDoc(
     ts.factory.createGetAccessorDeclaration(
-      undefined,
       undefined,
       getFieldName(fieldDescriptor),
       [],
@@ -1073,10 +1059,8 @@ function createOneOfGetter(
             ts.factory.createTypeLiteralNode([
               ts.factory.createIndexSignature(
                 undefined,
-                undefined,
                 [
                   ts.factory.createParameterDeclaration(
-                    undefined,
                     undefined,
                     undefined,
                     "index",
@@ -1116,7 +1100,6 @@ function createOneOfGetter(
 
   return ts.factory.createGetAccessorDeclaration(
     undefined,
-    undefined,
     oneofDescriptor.name,
     [],
     undefined,
@@ -1151,11 +1134,9 @@ function createSetter(
   return comment.addDeprecatedJsDoc(
     ts.factory.createSetAccessorDeclaration(
       undefined,
-      undefined,
       getFieldName(fieldDescriptor),
       [
         ts.factory.createParameterDeclaration(
-          undefined,
           undefined,
           undefined,
           valueParameter,
@@ -1255,7 +1236,6 @@ function createPresenceHas(
 ) {
   return comment.addDeprecatedJsDoc(
     ts.factory.createGetAccessorDeclaration(
-      undefined,
       undefined,
       getPrefixedFieldName("has", fieldDescriptor),
       [],
@@ -1471,7 +1451,6 @@ function createSerialize(
                 ts.factory.createParameterDeclaration(
                   undefined,
                   undefined,
-                  undefined,
                   "item",
                   undefined,
                   type.getTypeReference(
@@ -1613,7 +1592,6 @@ function createSerialize(
     ts.factory.createMethodDeclaration(
       undefined,
       undefined,
-      undefined,
       "serialize",
       undefined,
       undefined,
@@ -1624,13 +1602,11 @@ function createSerialize(
     ts.factory.createMethodDeclaration(
       undefined,
       undefined,
-      undefined,
       "serialize",
       undefined,
       undefined,
       [
         ts.factory.createParameterDeclaration(
-          undefined,
           undefined,
           undefined,
           "w",
@@ -1647,13 +1623,11 @@ function createSerialize(
     ts.factory.createMethodDeclaration(
       undefined,
       undefined,
-      undefined,
       "serialize",
       undefined,
       undefined,
       [
         ts.factory.createParameterDeclaration(
-          undefined,
           undefined,
           undefined,
           "w",
@@ -2054,7 +2028,6 @@ function createDeserialize(
   );
 
   return ts.factory.createMethodDeclaration(
-    undefined,
     [ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)],
     undefined,
     ts.factory.createIdentifier("deserialize"),
@@ -2062,7 +2035,6 @@ function createDeserialize(
     undefined,
     [
       ts.factory.createParameterDeclaration(
-        undefined,
         undefined,
         undefined,
         ts.factory.createIdentifier("bytes"),
@@ -2100,7 +2072,6 @@ function createDeserializeBinary(
     );
   }
   return ts.factory.createMethodDeclaration(
-    undefined,
     modifiers,
     undefined,
     ts.factory.createIdentifier("deserializeBinary"),
@@ -2108,7 +2079,6 @@ function createDeserializeBinary(
     undefined,
     [
       ts.factory.createParameterDeclaration(
-        undefined,
         undefined,
         undefined,
         ts.factory.createIdentifier("bytes"),
@@ -2142,7 +2112,6 @@ function createDeserializeBinary(
  */
 function createSerializeBinary(): ts.ClassElement {
   return ts.factory.createMethodDeclaration(
-    undefined,
     undefined,
     undefined,
     ts.factory.createIdentifier("serializeBinary"),
@@ -2193,7 +2162,6 @@ function createOneOfDecls(
     decls.push(ts.factory.createArrayLiteralExpression(map));
   }
   return ts.factory.createPropertyDeclaration(
-    undefined,
     [],
     ts.factory.createPrivateIdentifier("#one_of_decls"),
     undefined,
@@ -2263,7 +2231,6 @@ function createMessage(
   // Create message class
   return comment.addDeprecatedJsDoc(
     ts.factory.createClassDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
       `${parentName}${messageDescriptor.name}`,
       undefined,

--- a/src/index.ts
+++ b/src/index.ts
@@ -18,7 +18,6 @@ function createImport(
 ): ts.ImportDeclaration {
   return ts.factory.createImportDeclaration(
     undefined,
-    undefined,
     ts.factory.createImportClause(
       false,
       ts.factory.createNamespaceImport(identifier) as any,
@@ -34,7 +33,7 @@ function replaceExtension(filename: string, extension: string = ".ts"): string {
 
 
 const request = plugin.CodeGeneratorRequest.deserialize(
-  new Uint8Array(fs.readFileSync(0)),
+  new Uint8Array(fs.readFileSync(process.stdin.fd)),
 );
 const response = new plugin.CodeGeneratorResponse({
   supported_features:
@@ -42,13 +41,23 @@ const response = new plugin.CodeGeneratorResponse({
   file: [],
 });
 
+const [tsVersionMajor, tsVersionMinor] = ts.version.split(".").map(Number);
+
+if (!(tsVersionMajor >= 4 && tsVersionMinor >= 9)) {
+  response.error = "protoc-gen-ts requires TypeScript 4.9 or above.";
+  process.stdout.write(response.serialize());
+  // will help to flush the stdout buffer.
+  process.stdout.write(new Uint8Array());
+  process.exit();
+}
+
 const options = op.parse(request.parameter);
 
 type.initialize(options);
 descriptor.initialize(options);
 
-for (const descriptor of request.proto_file) {
-  type.preprocess(descriptor, descriptor.name, `.${descriptor.package ?? ""}`);
+for (const fileDescriptor of request.proto_file) {
+  type.preprocess(fileDescriptor, fileDescriptor.name, `.${fileDescriptor.package ?? ""}`);
 }
 
 for (const fileDescriptor of request.proto_file) {

--- a/src/rpc/node.ts
+++ b/src/rpc/node.ts
@@ -159,7 +159,6 @@ function createUnaryRpcPromiseMethod(
   return comment.addDeprecatedJsDoc(
     ts.factory.createPropertyDeclaration(
       undefined,
-      undefined,
       methodDescriptor.name,
       undefined,
       ts.factory.createTypeReferenceNode("GrpcPromiseServiceInterface", [
@@ -233,7 +232,6 @@ function createUnaryRpcMethod(
 
   return comment.addDeprecatedJsDoc(
     ts.factory.createPropertyDeclaration(
-      undefined,
       undefined,
       methodDescriptor.name,
       undefined,
@@ -342,7 +340,6 @@ function createClientStreamingRpcMethod(
   return comment.addDeprecatedJsDoc(
     ts.factory.createPropertyDeclaration(
       undefined,
-      undefined,
       methodDescriptor.name,
       undefined,
       ts.factory.createTypeReferenceNode("GrpWritableServiceInterface", [
@@ -434,7 +431,6 @@ function createServerStreamingRpcMethod(
   return comment.addDeprecatedJsDoc(
     ts.factory.createPropertyDeclaration(
       undefined,
-      undefined,
       methodDescriptor.name,
       undefined,
       ts.factory.createTypeReferenceNode("GrpcStreamServiceInterface", [
@@ -520,7 +516,6 @@ function createBidiStreamingRpcMethod(
   return comment.addDeprecatedJsDoc(
     ts.factory.createPropertyDeclaration(
       undefined,
-      undefined,
       methodDescriptor.name,
       undefined,
       ts.factory.createTypeReferenceNode("GrpcChunkServiceInterface", [
@@ -583,7 +578,6 @@ export function createServiceClient(
 ) {
   const members: Array<ts.ClassElement> = [
     ts.factory.createConstructorDeclaration(
-      undefined,
       undefined,
       [
         util.createParameter(
@@ -677,7 +671,6 @@ export function createServiceClient(
 
   return comment.addDeprecatedJsDoc(
     ts.factory.createClassDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
       ts.factory.createIdentifier(`${serviceDescriptor.name}Client`),
       undefined,

--- a/src/rpc/server.ts
+++ b/src/rpc/server.ts
@@ -249,11 +249,10 @@ function createServiceDefinition(
   // }
   const unaryIface = ts.factory.createInterfaceDeclaration(
     undefined,
-    undefined,
     "GrpcUnaryServiceInterface",
     [
-      ts.factory.createTypeParameterDeclaration("P"),
-      ts.factory.createTypeParameterDeclaration("R"),
+      ts.factory.createTypeParameterDeclaration(undefined, "P"),
+      ts.factory.createTypeParameterDeclaration(undefined, "R"),
     ],
     undefined,
     [
@@ -291,11 +290,10 @@ function createServiceDefinition(
   // }
   const promiseIface = ts.factory.createInterfaceDeclaration(
     undefined,
-    undefined,
     "GrpcPromiseServiceInterface",
     [
-      ts.factory.createTypeParameterDeclaration("P"),
-      ts.factory.createTypeParameterDeclaration("R"),
+      ts.factory.createTypeParameterDeclaration(undefined, "P"),
+      ts.factory.createTypeParameterDeclaration(undefined, "R"),
     ],
     undefined,
     [
@@ -318,11 +316,10 @@ function createServiceDefinition(
   // }
   const streamIface = ts.factory.createInterfaceDeclaration(
     undefined,
-    undefined,
     "GrpcStreamServiceInterface",
     [
-      ts.factory.createTypeParameterDeclaration("P"),
-      ts.factory.createTypeParameterDeclaration("R"),
+      ts.factory.createTypeParameterDeclaration(undefined, "P"),
+      ts.factory.createTypeParameterDeclaration(undefined, "R"),
     ],
     undefined,
     [
@@ -347,11 +344,10 @@ function createServiceDefinition(
   // }
   const writableIface = ts.factory.createInterfaceDeclaration(
     undefined,
-    undefined,
     "GrpWritableServiceInterface",
     [
-      ts.factory.createTypeParameterDeclaration("P"),
-      ts.factory.createTypeParameterDeclaration("R"),
+      ts.factory.createTypeParameterDeclaration(undefined, "P"),
+      ts.factory.createTypeParameterDeclaration(undefined, "R"),
     ],
     undefined,
     [
@@ -384,11 +380,10 @@ function createServiceDefinition(
   // }
   const chunkIface = ts.factory.createInterfaceDeclaration(
     undefined,
-    undefined,
     "GrpcChunkServiceInterface",
     [
-      ts.factory.createTypeParameterDeclaration("P"),
-      ts.factory.createTypeParameterDeclaration("R"),
+      ts.factory.createTypeParameterDeclaration(undefined, "P"),
+      ts.factory.createTypeParameterDeclaration(undefined, "R"),
     ],
     undefined,
     [
@@ -419,7 +414,6 @@ export function createUnimplementedServer(
 ) {
   const members: Array<ts.ClassElement> = [
     ts.factory.createPropertyDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.StaticKeyword)],
       "definition",
       undefined,
@@ -427,7 +421,6 @@ export function createUnimplementedServer(
       createServiceDefinition(rootDescriptor, serviceDescriptor),
     ),
     ts.factory.createIndexSignature(
-      undefined,
       undefined,
       [
         util.createParameter(
@@ -491,7 +484,6 @@ export function createUnimplementedServer(
 
     members.push(
       ts.factory.createMethodDeclaration(
-        undefined,
         [ts.factory.createModifier(ts.SyntaxKind.AbstractKeyword)],
         undefined,
         methodDescriptor.name,
@@ -506,7 +498,6 @@ export function createUnimplementedServer(
 
   return comment.addDeprecatedJsDoc(
     ts.factory.createClassDeclaration(
-      undefined,
       [
         ts.factory.createModifier(ts.SyntaxKind.ExportKeyword),
         ts.factory.createModifier(ts.SyntaxKind.AbstractKeyword),

--- a/src/rpc/util.ts
+++ b/src/rpc/util.ts
@@ -13,7 +13,6 @@ export function createParameter(
   return ts.factory.createParameterDeclaration(
     undefined,
     undefined,
-    undefined,
     name,
     optional ? ts.factory.createToken(ts.SyntaxKind.QuestionToken) : undefined,
     typename,

--- a/src/rpc/web.ts
+++ b/src/rpc/web.ts
@@ -52,7 +52,6 @@ function createMethodDescriptor(
   );
 
   return ts.factory.createPropertyDeclaration(
-    undefined,
     [
       ts.factory.createModifier(ts.SyntaxKind.PrivateKeyword),
       ts.factory.createModifier(ts.SyntaxKind.StaticKeyword),
@@ -80,7 +79,6 @@ function createUnaryRpcMethod(
     ts.factory.createTypeReferenceNode(ts.factory.createQualifiedName(grpcIdentifier, "Metadata")),
   );
   return ts.factory.createMethodDeclaration(
-    undefined,
     undefined,
     undefined,
     methodDescriptor.name,
@@ -176,7 +174,6 @@ function createUnaryRpcMethod(
   return ts.factory.createMethodDeclaration(
     undefined,
     undefined,
-    undefined,
     methodDescriptor.name,
     undefined,
     undefined,
@@ -254,7 +251,6 @@ function createServerStreamingRpcMethod(
   return ts.factory.createMethodDeclaration(
     undefined,
     undefined,
-    undefined,
     methodDescriptor.name,
     undefined,
     undefined,
@@ -326,7 +322,6 @@ export function createServiceClient(
 ) {
   const members: ts.ClassElement[] = [
     ts.factory.createPropertyDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.PrivateKeyword)],
       "_address",
       undefined,
@@ -334,7 +329,6 @@ export function createServiceClient(
       undefined,
     ),
     ts.factory.createPropertyDeclaration(
-      undefined,
       [ts.factory.createModifier(ts.SyntaxKind.PrivateKeyword)],
       "_client",
       undefined,
@@ -344,7 +338,6 @@ export function createServiceClient(
       undefined
     ),
     ts.factory.createConstructorDeclaration(
-      undefined,
       undefined,
       [
         util.createParameter(
@@ -479,7 +472,6 @@ export function createServiceClient(
   }
 
   return ts.factory.createClassDeclaration(
-    undefined,
     [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
     ts.factory.createIdentifier(`${serviceDescriptor.name}Client`),
     undefined,

--- a/test/conformance/maps/maps.ts
+++ b/test/conformance/maps/maps.ts
@@ -171,13 +171,13 @@ export namespace maps {
                 data.key = this.key;
             }
             if (this.keys != null) {
-                data.keys = Object.fromEntries(this.keys);
+                data.keys = (Object.fromEntries)(this.keys);
             }
             if (this.topics != null) {
-                data.topics = Object.fromEntries(Array.from(this.topics).map(([key, value]) => [key, value.toObject()]));
+                data.topics = (Object.fromEntries)((Array.from)(this.topics).map(([key, value]) => [key, value.toObject()]));
             }
             if (this.topics_with_intkeys != null) {
-                data.topics_with_intkeys = Object.fromEntries(Array.from(this.topics_with_intkeys).map(([key, value]) => [key, value.toObject()]));
+                data.topics_with_intkeys = (Object.fromEntries)((Array.from)(this.topics_with_intkeys).map(([key, value]) => [key, value.toObject()]));
             }
             return data;
         }

--- a/test/default/default_proto2.ts
+++ b/test/default/default_proto2.ts
@@ -435,10 +435,10 @@ export class DefaultMessageV2WithoutDefault extends pb_1.Message {
             data.int_but_string = this.int_but_string;
         }
         if (this.map_string_string != null) {
-            data.map_string_string = Object.fromEntries(this.map_string_string);
+            data.map_string_string = (Object.fromEntries)(this.map_string_string);
         }
         if (this.map_string_message != null) {
-            data.map_string_message = Object.fromEntries(Array.from(this.map_string_message).map(([key, value]) => [key, value.toObject()]));
+            data.map_string_message = (Object.fromEntries)((Array.from)(this.map_string_message).map(([key, value]) => [key, value.toObject()]));
         }
         if (this.array_message != null) {
             data.array_message = this.array_message.map((item: dependency_1.DefaultCommonMessage) => item.toObject());

--- a/test/default/default_proto3.ts
+++ b/test/default/default_proto3.ts
@@ -469,10 +469,10 @@ export class DefaultMessageV3 extends pb_1.Message {
             data.int_but_string = this.int_but_string;
         }
         if (this.map_string_string != null) {
-            data.map_string_string = Object.fromEntries(this.map_string_string);
+            data.map_string_string = (Object.fromEntries)(this.map_string_string);
         }
         if (this.map_string_message != null) {
-            data.map_string_message = Object.fromEntries(Array.from(this.map_string_message).map(([key, value]) => [key, value.toObject()]));
+            data.map_string_message = (Object.fromEntries)((Array.from)(this.map_string_message).map(([key, value]) => [key, value.toObject()]));
         }
         if (this.array_int32 != null) {
             data.array_int32 = this.array_int32;

--- a/test/map.ts
+++ b/test/map.ts
@@ -192,16 +192,16 @@ export class Tags extends pb_1.Message {
             data.key = this.key;
         }
         if (this.keys != null) {
-            data.keys = Object.fromEntries(this.keys);
+            data.keys = (Object.fromEntries)(this.keys);
         }
         if (this.topics != null) {
-            data.topics = Object.fromEntries(Array.from(this.topics).map(([key, value]) => [key, value.toObject()]));
+            data.topics = (Object.fromEntries)((Array.from)(this.topics).map(([key, value]) => [key, value.toObject()]));
         }
         if (this.imported != null) {
-            data.imported = Object.fromEntries(Array.from(this.imported).map(([key, value]) => [key, value.toObject()]));
+            data.imported = (Object.fromEntries)((Array.from)(this.imported).map(([key, value]) => [key, value.toObject()]));
         }
         if (this.imported2 != null) {
-            data.imported2 = Object.fromEntries(this.imported2);
+            data.imported2 = (Object.fromEntries)(this.imported2);
         }
         return data;
     }

--- a/test/presence/presence_proto2.ts
+++ b/test/presence/presence_proto2.ts
@@ -244,7 +244,7 @@ export class PresenceMessageV2 extends pb_1.Message {
             data.opt_oneof = this.opt_oneof.toObject();
         }
         if (this.map != null) {
-            data.map = Object.fromEntries(this.map);
+            data.map = (Object.fromEntries)(this.map);
         }
         return data;
     }

--- a/test/presence/presence_proto3.ts
+++ b/test/presence/presence_proto3.ts
@@ -313,7 +313,7 @@ export class PresenceMessageV3 extends pb_1.Message {
             data.repeated = this.repeated;
         }
         if (this.map != null) {
-            data.map = Object.fromEntries(this.map);
+            data.map = (Object.fromEntries)(this.map);
         }
         return data;
     }

